diff --git a/target/sim/sw/device/Makefile b/target/sim/sw/device/Makefile
index 1bab68e..43a8b30 100644
--- a/target/sim/sw/device/Makefile
+++ b/target/sim/sw/device/Makefile
@@ -7,6 +7,7 @@
 # Add user applications to APPS variable
 APPS  = blas/axpy
 APPS += blas/gemm
+APPS += blas/test
 
 TARGET ?= all
 
diff --git a/target/sim/sw/host/apps/offload/Makefile b/target/sim/sw/host/apps/offload/Makefile
index 990cc15..fe48749 100644
--- a/target/sim/sw/host/apps/offload/Makefile
+++ b/target/sim/sw/host/apps/offload/Makefile
@@ -18,6 +18,7 @@ APP  = offload
 SRCS = src/offload.c
 DEVICE_APPS  = blas/axpy
 DEVICE_APPS += blas/gemm
+DEVICE_APPS += blas/test
 
 # Compiler toolchain
 RISCV_CC      = riscv64-unknown-elf-gcc
